### PR TITLE
postgresql_database: Reassign objects owners if database owner changes

### PR DESCRIPTION
**Pull request by lukaalba**
Hi,

this is my first Pull Request, so if something is missing or wrong, let me know.

To resolve [439](https://github.com/cyrilgdn/terraform-provider-postgresql/issues/439) I suggest to add an additional parameter to the database resource called `alter_object_ownership`. If set to true, it will issue a REASSIGN OWNED BY query on the database when the owner changes, transfer the whole ownership in the database from the previous owner to the new one.

As far as I can tell this is only possible if the user in the provider configuration is a superuser or a user which has grants for both the previous owner role and the new one. So I skipped the test in the rds-like test suite.

### CODE REVIEW

File: website/docs/r/postgresql_database.html.markdown [hunk 1]
```diff
@@ -82,6 +83,14 @@ resource "postgresql_database" "my_db" {
   force the creation of a new resource as this value can only be changed when a
   database is created.
 
+* `alter_object_ownership` - (Optional) If `true`, the change of the database
+  `owner` will also include a reassignment of the ownership of preexisting
+  objects like tables or sequences from the previous owner to the new one.
+  If set to `false` (the default), then the previous database `owner` will still
+  hold the ownership of the objects in that database. To alter existing objects in
+  the database, you must be a direct or indirect member of the specified role, or
+  the username in the provider is a superuser.
```
Comment (cyrilgdn) [inline website/docs/r/postgresql_database.html.markdown:92]:
```suggestion
  the username in the provider must be superuser.
```

File: postgresql/resource_postgresql_database.go [hunk 2]
```diff
@@ -476,6 +488,54 @@ func setDBOwner(db *DBConnection, d *schema.ResourceData) error {
 	return err
 }
 
+func setAlterOwnership(db *DBConnection, d *schema.ResourceData) error {
+	if d.HasChange(dbOwnerAttr) || d.HasChange(dbAlterObjectOwnership) {
```
Comment (cyrilgdn) [inline postgresql/resource_postgresql_database.go:492]: if only `dbAlterObjectOwnership` changes, it seems you will execute the `REASSIGN` with the same previous and new owner so I'm not sure to see the point  on this case :thinking: 

At least you should check if `currentOwner` and `newOwner` are different below.

**Review by cyrilgdn — COMMENTED** [review]
Hi @lukaalba,

Thanks for your work on that :muscle: (and to add the acceptance tests :pray: )

I've just allowed myself to push a [small modification](https://github.com/cyrilgdn/terraform-provider-postgresql/pull/458/files#diff-a53ac37ef5a228c8bc85a0cb082f0e58826c351ae96ffc6c940e8df3a2e71056R294-R296) in the tests to drop the role you created after each test so we can execute the same tests multiple times (without recreating the whole database server)

I have also one question/comment on the main function.

But otherwise it seems good to me :+1:

File: postgresql/resource_postgresql_database.go [hunk 3]
```diff
@@ -393,6 +400,10 @@ func resourcePostgreSQLDatabaseUpdate(db *DBConnection, d *schema.ResourceData)
 		return err
 	}
 
+	if err := setAlterOwnership(db, d); err != nil {
```
Comment (cyrilgdn) [inline postgresql/resource_postgresql_database.go:403]: This does more than setting a database "parameter", probably this function could have better name
(something like `reassignOwnership` or similar)

File: postgresql/resource_postgresql_database.go (hunk 2 above)
Comment (lukaalba) [reply postgresql/resource_postgresql_database.go:492]: Thanks for the review. If updated the code so it checks if currentOwner and newOwner are equal before executing the sql statements

File: postgresql/resource_postgresql_database.go [hunk 4]
```diff
@@ -476,6 +488,57 @@ func setDBOwner(db *DBConnection, d *schema.ResourceData) error {
 	return err
 }
 
+func setAlterOwnership(db *DBConnection, d *schema.ResourceData) error {
+	if d.HasChange(dbOwnerAttr) || d.HasChange(dbAlterObjectOwnership) {
```
[the remark this answers is not in the document]
Comment (lukaalba) [reply postgresql/resource_postgresql_database.go]: I agree with this one. I've changed it in my latest commit

**Comment by cyrilgdn** [comment]
I've just fix the used of the transaction (which fixes at the same time the `golanci-lint` issue), we prefer to `defer` the rollback and commit, otherwise it will try to commit in any cases .

**Review by cyrilgdn — APPROVED** [review]

**Review by cyrilgdn — APPROVED** [review]

**Comment by cyrilgdn** [comment]
@lukaalba This has just been released in [v1.23.0](https://github.com/cyrilgdn/terraform-provider-postgresql/releases/tag/v1.23.0)

**Pull request merged**